### ISSUE TITLE
ONNX model produces different results in tensorRT 7 compared to onnx runtime and tensorRT 6

### ISSUE THREAD

**Issue by ttdd11** [NVIDIA/TensorRT#974]
Differences in output values when comparing results from trt 7.2.1 to onnx runtime and trt 6. 

TensorRT Version: 7.2.1
GPU Type: 3080
Nvidia Driver Version: 460.89
CUDA Version: 11.1
CUDNN Version: 8.0
Operating System:  Windows 10

For this test, the model was developed in tensorflow and converted to onnx using the latest tf2onnx. I tried using opset 12,11, 7, and all produced comparable results in onnx runtime and trt 6, but unfortunately the results look strange in trt 7. All results seem to be shifted. 

Edit: I think the upsample layer is causing the problems here. So, I tried onnx-simplifier and this removed the upsample for resize, however the results are the same. One warning is this:
 [12/21/2020-07:34:45] [W] [TRT] onnx2trt_utils.cpp:220: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.

Simplified onnx model also produces the same results in onnx runtime.

It's hard to imagine that this cast is causing the problems here, but I may be wrong.

Can provide the  onnx runtime sample script and onnx file.

**Comment by ttyio** [comment]
Hello @ttdd11 , thanks for reporting.
Could you share your ONNX model? thanks!

**Comment by ttdd11** [comment]
Do you have an email I can send to?

**Comment by ttyio** [comment]
any cloud storage fine to me, you can also send ttyio#hotmail.com, thanks

**Comment by ttdd11** [comment]
Just sent. Can you reply here to confirm that you have received the files?

**Comment by ttyio** [comment]
received thanks @ttdd11

**Comment by ttdd11** [comment]
@ttyio thank you. Let me know if you need anything else.

**Comment by ttyio** [comment]
Hello @ttdd11 , model received, but there is no windows machine I can try today.
I checked the model and did see obvious limitation, could you help to try:
- check if the mismatch come from TF32 (https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#tf32-inference-c)
- try TRT7.2 + CUDA10.2 on Windows? 
- try Linux if it is possible?

**Comment by ttdd11** [comment]
@ttyio thanks for taking a look.

1. Removed the fp16 flag, network built however the results are similar.
2. Tried TRT 7.2.1 cuda 10.2 cudnn 8 with this error:
[F] [TRT] Assertion failed: (smVersion < SM_VERSION_A100) && "SM version not supported in this NVRTC version"
and
[E] [TRT] C:/source/rtExt\helpers.h (523) - Assertion Error in nvinfer1::fixSMVerForNVRTC: 0 ((smVersion < SM_VERSION_A100) && "SM version not supported in this NVRTC version")

This makes sense to me seeing as this is a 3080 which I think requires cuda 11.

3. Linux is a bit challenging for me to test.

Is there anything else I can do?

**Comment by ttyio** [comment]
Hello @ttdd11 , thanks for the experiment,
Which metric did you use when you check the output difference? Did you  the same input and the same ONNX model when you test TRT6 result vs. onnxruntime? Also have you tried onnxruntime-gpu or tf-onnx run on gpu? thanks!

**Comment by ttdd11** [comment]
@ttyio thanks for the continued help. Currently the test is somewhat qualitative because they aren't even close. For onnxruntime versus trt6, the results are within a very fine tolerance. With trt7, the values are shifted an amount that is visually obvious. 

For all tests, the same onnx model is being used. I just tried onnxruntime-gpu and the results are comparable to the cpu, trt6, but not trt 7.

I tried using polygraphy, however this is linux based and there are no python bindings for trt on windows so I was unable to finish this test.

Were you able to verify differences for the model using onnxruntime versus trt 7, or using polygraphy on this onnx model on a linux machine?

Let me know what else I can do.

**Comment by ttyio** [comment]
Hello @ttdd11 , thanks for your input. I get Windows machine today, here is the finding:
- the Windows output match Linux output in TRT
- the TRT7 output mismatch ONNXruntime/TRT6 output because of Resize/Upsample

More details:
The mismatch start from the `Upsample` whose input activation is `tower_0/res5/refine_conv2/Relu:0`. 

In TRT6, we use the same coordinate transformation in resize for both linear and NN, and during development of TRT7, we change the implementation to follow pytorch, and now we use different coordinate transformation for NN and linear. The `asymmetric` exported from TF is broken because of this change (we use pytorch_half_pixel, that's the reason you see the output is shifted;-( ).

The fix is scheduled in next major release. Before that, TF resize `linear` + `alignCorner = 0` + `halfPixel = 1` works in TRT7.x, or can you use back TRT6 before next major release? sorry for the bug ;-(

**Comment by ttdd11** [comment]
@ttyio thank you for the very detailed report.

In terms of the fix, this is somewhat time sensitive. Would you mind elaborating on the fix you describe and how I would implement that?

**Comment by ttdd11** [comment]
@ttyio just to follow up on my previous post. Is the fix to change this:

tf.compat.v1.image.resize_bilinear(
    images, size, align_corners=False, name=None, half_pixel_centers=False
)

With align corners to be false and half_pixel_centers to be true in all resize functions, and then re-export the .pb from tf and then convert to onnx? Is there a way to fix this in the onnx file?

Thanks again.

**Comment by ttdd11** [comment]
@ttyio Sorry for all the messages.

I naively added half_pixel_centers to true in tf and and seeing the shifting present in trt7.

Now the question is, is there a work around for this before the next major release? Is there anything I can do in Tensorflow/ONNX/TRT7 to account for this? If I build the OSS components from source will this correct the problem?

If no, if there a tentative approximate release date for this push?

**Comment by ttyio** [comment]
Sorry @ttdd11 for the delay response , 

First I need correct myself, I just verified when use `linear` + `alignCorner = 1` + `halfPixel = 0`,  the exported onnx has coordinate transformation  `align_corners`, TF, onnxruntime and TRT7.x output the same value in this configuration.

Because `align_corners` formula is different than the `asymmetric` in TRT6, so you need fine-tune your model using `linear` + `alignCorner = 1` + `halfPixel = 0`. Then export to ONNX and run with TRT.

The support of resize is inside TRT core and not opensourced yet, so sorry we cannot workaround it by fix and rebuild OSS.

And for the question about approximate release date for next major release, sorry I have no answer here.

**Comment by ttdd11** [comment]
@ttyio Thanks for the response. 

The tune is no problem, the question I have is regarding the tensorflow version. I was reading another support ticket where the resize has a different implementation until tf > 2.0 due to a ceil operator. Is it true that I will need to run this in tf > 2.0 for this process to work or is 1.14/1.15 okay?

I realize this is probably not your purview, however I believe you were responding to the ticket I am referring to that I can't seem to find now.

**Comment by ttyio** [comment]
Hello @ttdd11 . From my understanding, the fix is in https://github.com/tensorflow/tensorflow/commit/3ae2c6691b7c6e0986d97b150c9283e5cc52c15f, they added new attribute `half_pixel_centers`. So as long as we have this commit, It is fine to set `half_pixel_centers = 0`, `align_corners = 1` and run the training. Thanks~

**Comment by ttdd11** [comment]
@ttdd11 suggestions work. Slight changes in output but the network is fundamentally different after the tune. Will train with more similar preferences but as far as I can see this is resolved. Thanks for the help.

**Comment by Egozjuer** [comment]
> Sorry @ttdd11 for the delay response ,
> 
> First I need correct myself, I just verified when use `linear` + `alignCorner = 1` + `halfPixel = 0`, the exported onnx has coordinate transformation `align_corners`, TF, onnxruntime and TRT7.x output the same value in this configuration.
> 
> Because `align_corners` formula is different than the `asymmetric` in TRT6, so you need fine-tune your model using `linear` + `alignCorner = 1` + `halfPixel = 0`. Then export to ONNX and run with TRT. Is this workable for you?
> 
> The support of resize is inside TRT core and not opensourced yet, so sorry we cannot workaround it by fix and rebuild OSS.
> 
> And for the question about approximate release date for next major release, sorry I have no answer here.

@ttyio hi，my onnx resizes ops has both two formula,"align_corners" and "asymmetric" ,did you mean "align_corners" is correct but " "asymmetric" is wrong in TRT7?

**Comment by ttyio** [comment]
Hi @Egozjuer , 
TF fixed some resize issue in TF2, halfPixel now used for compatibility between TF1 and TF2, and use the combination of half_pixel_centers  and align_corners, we can get 4 different formulas. And these formulas are not the same as in Pytorch.

TRT 7.x only has one param `alignCorners`, so we cannot support both TF and Pytorch. we only support half_pixel_centers = 0, align_corners = 1 for `linear` resize. So could you either change your code to this supported combination, or wait for next major release? thanks!

**Comment by batselem** [comment]
@ttyio, I am having the same issue.

```
Assertion failed: (smVersion < SM_VERSION_A100) && "SM version not supported in this NVRTC version"
C:/source/rtExt\helpers.h:523
Aborting...
[02/01/2021-11:16:51] [E] [TRT] C:/source/rtExt\helpers.h (523) - Assertion Error in nvinfer1::fixSMVerForNVRTC: 0 ((smVersion < SM_VERSION_A100) && "SM version not supported in this NVRTC version")
[02/01/2021-11:16:51] [E] Engine creation failed
```
This error occurred after I switch RTX 4000 to RTX 3070. I tried to build TRT on windows for RTX 3070 with GPU_ARCH=86. But your windows instructions didn't work unfortunately. When is next major TRT release? I wonder it will include support for compute capability 8.6.

**Comment by ttyio** [comment]
Hello @batselem , could you switch to CUDA 11.0+ and retry? thanks

**Comment by batselem** [comment]
But I am using TensorRT-7.2.1.6.Windows10.x86_64.cuda-11.0.cudnn8.0. Here is my issue post. Thank you for your reply.
https://github.com/NVIDIA/TensorRT/issues/1035

**Comment by batselem** [comment]
Here is the error message:
```
[02/01/2021-11:42:04] [I] === Model Options ===
[02/01/2021-11:42:04] [I] Format: ONNX
[02/01/2021-11:42:04] [I] Model: C:/Users/fodics/Downloads/yolov4.onnx
[02/01/2021-11:42:04] [I] Output:
[02/01/2021-11:42:04] [I] === Build Options ===
[02/01/2021-11:42:04] [I] Max batch: explicit
[02/01/2021-11:42:04] [I] Workspace: 16 MiB
[02/01/2021-11:42:04] [I] minTiming: 1
[02/01/2021-11:42:04] [I] avgTiming: 8
[02/01/2021-11:42:04] [I] Precision: FP32
[02/01/2021-11:42:04] [I] Calibration:
[02/01/2021-11:42:04] [I] Refit: Disabled
[02/01/2021-11:42:04] [I] Safe mode: Disabled
[02/01/2021-11:42:04] [I] Save engine: yolo4v.engine
[02/01/2021-11:42:04] [I] Load engine:
[02/01/2021-11:42:04] [I] Builder Cache: Enabled
[02/01/2021-11:42:04] [I] NVTX verbosity: 0
[02/01/2021-11:42:04] [I] Tactic sources: Using default tactic sources
[02/01/2021-11:42:04] [I] Input(s)s format: fp32:CHW
[02/01/2021-11:42:04] [I] Output(s)s format: fp32:CHW
[02/01/2021-11:42:04] [I] Input build shapes: model
[02/01/2021-11:42:04] [I] Input calibration shapes: model
[02/01/2021-11:42:04] [I] === System Options ===
[02/01/2021-11:42:04] [I] Device: 0
[02/01/2021-11:42:04] [I] DLACore:
[02/01/2021-11:42:04] [I] Plugins:
[02/01/2021-11:42:04] [I] === Inference Options ===
[02/01/2021-11:42:04] [I] Batch: Explicit
[02/01/2021-11:42:04] [I] Input inference shapes: model
[02/01/2021-11:42:04] [I] Iterations: 10
[02/01/2021-11:42:04] [I] Duration: 3s (+ 200ms warm up)
[02/01/2021-11:42:04] [I] Sleep time: 0ms
[02/01/2021-11:42:04] [I] Streams: 1
[02/01/2021-11:42:04] [I] ExposeDMA: Disabled
[02/01/2021-11:42:04] [I] Data transfers: Enabled
[02/01/2021-11:42:04] [I] Dump refittable layers:Disabled
[02/01/2021-11:42:04] [I] Dump output: Disabled
[02/01/2021-11:42:04] [I] Profile: Disabled
[02/01/2021-11:42:04] [I] Export timing to JSON file:
[02/01/2021-11:42:04] [I] Export output to JSON file:
[02/01/2021-11:42:04] [I] Export profile to JSON file:
[02/01/2021-11:42:04] [I]
[02/01/2021-11:42:04] [I] === Device Information ===
[02/01/2021-11:42:04] [I] Selected Device: GeForce RTX 3070
[02/01/2021-11:42:04] [I] Compute Capability: 8.6
[02/01/2021-11:42:04] [I] SMs: 46
[02/01/2021-11:42:04] [I] Compute Clock Rate: 1.725 GHz
[02/01/2021-11:42:04] [I] Device Global Memory: 8192 MiB
[02/01/2021-11:42:04] [I] Shared Memory per SM: 100 KiB
[02/01/2021-11:42:04] [I] Memory Bus Width: 256 bits (ECC disabled)
[02/01/2021-11:42:04] [I] Memory Clock Rate: 7.001 GHz
[02/01/2021-11:42:04] [I]
----------------------------------------------------------------
Input filename:   C:/Users/fodics/Downloads/yolov4.onnx
ONNX IR version:  0.0.6
Opset version:    11
Producer name:    tf2onnx
Producer version: 1.7.0
Domain:
Model version:    0
Doc string:
----------------------------------------------------------------
[02/01/2021-11:43:38] [W] [TRT] onnx2trt_utils.cpp:220: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
[02/01/2021-11:43:38] [W] Dynamic dimensions required for input: input_1:0, but no shapes were provided. Automatically overriding shape to: 1x416x416x3
[02/01/2021-11:44:27] [W] [TRT] TensorRT was linked against cuBLAS/cuBLAS LT 10.2.2 but loaded cuBLAS/cuBLAS LT 10.1.0
[02/01/2021-11:44:28] [I] [TRT] Some tactics do not have sufficient workspace memory to run. Increasing workspace size may increase performance, please check verbose output.
[02/01/2021-11:44:29] [F] [TRT] Assertion failed: (smVersion < SM_VERSION_A100) && "SM version not supported in this NVRTC version"
C:/source/rtExt\helpers.h:523
Aborting...
[02/01/2021-11:44:29] [E] [TRT] C:/source/rtExt\helpers.h (523) - Assertion Error in nvinfer1::fixSMVerForNVRTC: 0 ((smVersion < SM_VERSION_A100) && "SM version not supported in this NVRTC version")
[02/01/2021-11:44:29] [E] Engine creation failed
[02/01/2021-11:44:29] [E] Engine set up failed
```

**Comment by batselem** [comment]
I tested trtexec.exe with mnist.onnx cnn model on this environment setting and it just worked fine and the engine was successfully built but when I try YOLOv4.onnx model, it throws the above error.

**Comment by batselem** [comment]
@ttyio
I even tried creating my own builder function without using trtexec:
```
void buildEngine(const std::string &onnxFile, int maxBatchSzie, TRTUniquePtr<nvinfer1::ICudaEngine>& engine, TRTUniquePtr< nvinfer1::IExecutionContext >& context) {
	cudaSetDevice(0);
	
	auto builder = TRTUniquePtr<nvinfer1::IBuilder>(nvinfer1::createInferBuilder(gLogger));
	assert(builder != nullptr);
	
	const auto explicitBatch = 1U << static_cast<uint32_t>(nvinfer1::NetworkDefinitionCreationFlag::kEXPLICIT_BATCH);
	nvinfer1::INetworkDefinition* network = builder->createNetworkV2(explicitBatch);
	
	assert(network != nullptr);
	
	auto parser = TRTUniquePtr<nvonnxparser::IParser>(nvonnxparser::createParser(*network, gLogger));
	assert(parser != nullptr);

	if (!parser->parseFromFile(onnxFile.c_str(), 2))
	{
		std::string msg("failed to parse onnx file");
		gLogger.log(nvinfer1::ILogger::Severity::kERROR, msg.c_str());
		exit(EXIT_FAILURE);
	}

	builder->setMaxBatchSize(maxBatchSzie);
	builder->setMaxWorkspaceSize(1 << 30);// 1G

	std::cout << "setFp16Mode" << std::endl;
	if (!builder->platformHasFastFp16())
		std::cout << "Notice: the platform do not has fast for fp16" << std::endl;
	builder->setFp16Mode(true);

	std::cout << "Begin building engine..." << std::endl;
	engine = TRTUniquePtr<nvinfer1::ICudaEngine>(builder->buildCudaEngine(*network));

	if (!engine) {
		std::string error_message = "Unable to create engine";
		gLogger.log(nvinfer1::ILogger::Severity::kERROR, error_message.c_str());
		exit(-1);
	}
	std::cout << "End building engine..." << std::endl;
	
	context = TRTUniquePtr<nvinfer1::IExecutionContext>(engine->createExecutionContext());
	assert(context);

}
```

but still got the following error:
```
UNKNOWN: Fastest Tactic: 0 Time: 0.070656
UNKNOWN: --------------- Timing Runner: <reformat> (Reformat)
UNKNOWN: Tactic: 1002 time 0.192512
UNKNOWN: Tactic: 0 time 0.074752
UNKNOWN: Fastest Tactic: 0 Time: 0.074752
UNKNOWN: *************** Autotuning format combination: Float(1,416,173056,5537792) -> Float(1,416,173056,5537792) ***************
INTERNAL_ERROR: Assertion failed: (smVersion < SM_VERSION_A100) && "SM version not supported in this NVRTC version"
C:/source/rtExt\helpers.h:523
Aborting...
UNKNOWN: Builder timing cache: created 53 entries, 15 hit(s)
ERROR: C:/source/rtExt\helpers.h (523) - Assertion Error in nvinfer1::fixSMVerForNVRTC: 0 ((smVersion < SM_VERSION_A100) && "SM version not supported in this NVRTC version")
ERROR: Unable to create engine
```

**Comment by ttyio** [comment]
Hello @batselem , sorry I missed this thread,
Have you tried latest 7.2 release, also I see you are on windows, could you try dependency walker(https://www.dependencywalker.com/) to see which nvrtc.dll is loaded when you launch trtexec. thanks!
